File: 프로그래머스/lv3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    TreeSet<Integer> treeSet = new TreeSet<>();
    
    public void operate(String operation, Integer number){
        if(operation.equals("I")){
            treeSet.add(number);
        }
        else if(operation.equals("D")){
            if(!treeSet.isEmpty()){
                if(number == 1){
                    treeSet.pollLast();
                }
                else if(number == -1){
                    treeSet.pollFirst();
                }
            }
        }
    }
    
    public int[] solution(String[] operations) {
        
        for(String operation : operations){
            String[] _operation = operation.split(" ");
            String operator = _operation[0];
            Integer number = Integer.parseInt(_operation[1]);
            
            operate(operator, number);
        }
        
        
        
        
        
        return treeSet.isEmpty() ? new int[]{0,0} : new int[]{treeSet.last(), treeSet.first()};
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    maxQueue = []
    minQueue = []
    
    for operation in operations:
        operator, number = operation.split(" ")
        
        if(operator == 'I'):
            heapq.heappush(maxQueue, -int(number))
            heapq.heappush(minQueue, int(number))
        elif(operator == 'D' and minQueue):
            if(number == '1'): #최댓값 삭제
                heapq.heappop(maxQueue)
                minQueue = [-x for x in maxQueue]
                heapq.heapify(minQueue)
            elif(number == '-1'): #최솟값 삭제
                heapq.heappop(minQueue)
                maxQueue = [-x for x in minQueue]
                heapq.heapify(maxQueue)
    
    if not maxQueue:
        return [0,0]
    answer = [-maxQueue[0], minQueue[0]]
    return answer